src/dummy_slack_thread.jl

```diff
@@ -0,0 +1,54 @@
+"""
+    SlackCallRecord
+
+This object has fields
+
+* `call::Symbol`: either `:DummyThread` or `:send_exception_message`
+* `args::Tuple`: the arguments to the call
+* `kwargs::NamedTuple`: any keyword arguments to the call
+
+corresponding to function calls made with [`DummyThread`](@ref)s.
+"""
+struct SlackCallRecord
+    call::Symbol
+    args::Tuple
+    kwargs::NamedTuple
+end
+
+StructTypes.StructType(::Type{SlackCallRecord}) = StructTypes.Struct()
+# This makes it so JSON3 doesn't omit tuples/namedtuples if they are empty
+@inline StructTypes.isempty(::Type{SlackCallRecord}, ::NamedTuple) = false
+@inline StructTypes.isempty(::Type{SlackCallRecord}, ::Tuple) = false
+
+"""
+    DummyThread <: AbstractSlackThread
+
+Provides a "dummy" SlackThread which does not make any API calls
+to Slack, and instead logs messages to it's `logged` field.
+
+This can be used for testing or to pass to code expecting an `AbstractSlackThread`
+when you don't want to log anything to Slack.
+
+The `logged` field is a `Vector{SlackCallRecord}` corresponding to logging calls.
+
+See also: [`SlackCallRecord`](@ref).
+"""
+mutable struct DummyThread <: AbstractSlackThread
+    channel::Union{String,Nothing}
+    ts::Union{String,Nothing}
```

Comment:
In what scenario would these be mutated to be anything else?
```suggestion
    channel::Nothing
    ts::Nothing
```

Reply:
I have private code that sets the channel and thread from an object deserialized from JSON, and I wanted that to still work with a DummyThread too. I actually added that `Base.copyto!` method to provide a public API for this but ended up still doing field access since I wanted to set them separately at some point.

Maybe we can add public `set_channel!` and `set_thread!` that can be no-ops for `DummyThread`s?

Reply:
I'll point out there is an alternative where you just set `Base.setproperty` and `Base.getproperty` for `DummyThread` and remove these fields entirely

Reply:
Good call, switched to that.